homeassistant/components/media_player/sonos.py

```diff
@@ -682,6 +644,60 @@ def _refresh_music(self, update_media_position, track_info):
 
         self._source_name = None
 
+    def update_volume(self, event=None):
+        """Update information about currently volume settings."""
+        if event:
+            variables = event.variables
+
+            if 'volume' in variables:
+                self._player_volume = int(variables['volume']['Master'])
+
+            if 'mute' in variables:
+                self._player_muted = (variables['mute']['Master'] == '1')
+
+            if 'night_mode' in variables:
+                self._night_sound = (variables['night_mode'] == '1')
+
+            if 'dialog_level' in variables:
+                self._speech_enhance = (variables['dialog_level'] == '1')
+
+            self.schedule_update_ha_state()
+        else:
+            self._player_volume = self.soco.volume
+            self._player_muted = self.soco.mute
+            self._night_sound = self.soco.night_mode
+            self._speech_enhance = self.soco.dialog_mode
+
+    def update_groups(self, event=None):
+        """Process a zone group topology event coming from a player."""
+        if event:
+            self._receives_events = True
```

Comment:
Maybe for a future PR, would it make sense to automatically mark this as False when an event has not been received for X time?

Reply:
No events are received when a player is idle. Deciding between an idle player and an active player where we missed the playing event could be tricky. I hope that this complexity is not needed ... but yes, periodically resetting the variable could make sense if issues about stability do appear.